### PR TITLE
Week 2

### PR DESCRIPTION
**Pull request by YuRihl**
Зроблена аутентифікація, при логіні генерація JWT-токену, який перевіряє JwtStrategy, а JwtGuard захищає певні ендпоінти від доступу без токена (або неправильний токен, або час його дії минув)

### CODE REVIEW

File: src/user/user.controller.ts [hunk 1]
```diff
@@ -1,52 +1,53 @@
...
-  @Delete(':id')
-  remove(@Param('id') id: string) {
-    return this.userService.remove(+id);
+  @ApiOkResponse({
+    description: 'The user was deleted successfully',
+    type: User,
+  })
+  @UseGuards(JwtGuard)
+  @Delete('profile')
+  remove(@UserDecorator('id') id: number) {
```
Comment (bserefaniuk) [inline src/user/user.controller.ts:50]: So.. any user can delete any user?

File: src/user/user.service.ts [hunk 2]
```diff
@@ -1,26 +1,44 @@
...
 
-  remove(id: number) {
-    return `This action removes a #${id} user`;
+  async remove(id: number) {
+    const user = await this.userRepository.findOneBy({ id });
+    const deletedUser = await this.userRepository.remove(user);
+
+    delete deletedUser.password;
+
+    return deletedUser;
```
Comment (bserefaniuk) [inline src/user/user.service.ts:42]: We should use a DTO object for this purpose

File: src/user/user.service.ts [hunk 3]
```diff
@@ -1,26 +1,44 @@
...
 
-  findOne(id: number) {
-    return `This action returns a #${id} user`;
-  }
+  async update(id: number, updateUserDto: UpdateUserDto) {
+    let user = await this.userRepository.findOneBy({ id });
+
+    const hashSalt = await bcrypt.genSalt();
+
+    user = {
```
Comment (bserefaniuk) [inline src/user/user.service.ts:23]: what if we want to update only one field?

File: src/user/user.controller.ts (hunk 1 above)
Comment (YuRihl) [reply src/user/user.controller.ts:50]: @UserDecorator('id') takes id from current user using JWT-token, so user can only delete himself

File: src/user/user.service.ts (hunk 2 above)
Comment (YuRihl) [reply src/user/user.service.ts:42]: I've added DTO for returning user but it doesn't filter object and doesn't delete password property as well

File: src/user/user.service.ts (hunk 3 above)
Comment (YuRihl) [reply src/user/user.service.ts:23]: UpdateDto has optional properties now so we can change any property of user using Patch

File: src/user/user.controller.ts [hunk 4]
```diff
@@ -1,52 +1,53 @@
...
-  @Patch(':id')
-  update(@Param('id') id: string, @Body() updateUserDto: UpdateUserDto) {
-    return this.userService.update(+id, updateUserDto);
+  @UseGuards(JwtGuard)
+  @Patch('profile')
+  update(
+    @UserDecorator('id') id: number,
+    @Body() updateUserDto: UpdateUserDto,
+  ) {
+    return this.userService.update(id, updateUserDto);
```
[the remark this answers is not in the document]
Comment (YuRihl) [reply src/user/user.controller.ts:41]: The solution is the same as deleting user

File: src/user/user.service.ts [hunk 5]
```diff
@@ -1,26 +1,44 @@
...
-    return 'This action adds a new user';
-  }
+  constructor(
+    @InjectRepository(User) private userRepository: Repository<User>,
+  ) {}
 
-  findAll() {
-    return `This action returns all user`;
+  findOne(user: User) {
+    return user;
```
[the remark this answers is not in the document]
Comment (YuRihl) [reply src/user/user.service.ts:15]: Done

File: src/auth/auth.service.ts [hunk 6]
```diff
@@ -1,4 +1,73 @@
...
+        secondName: userDto.secondName,
+      });
+
+      await this.userRepository.save(userToCreate);
+
+      const createdUser: User = await this.userRepository.findOneBy({
+        email: userDto.email,
+      });
+
+      console.log(createdUser);
```
[the remark this answers is not in the document]
Comment (YuRihl) [reply src/auth/auth.service.ts:55]: Done

File: src/auth/auth.service.ts [hunk 7]
```diff
@@ -1,4 +1,73 @@
...
+
+      return this.signToken(createdUser.email, createdUser.id);
+    } catch (error) {
+      console.error(error);
```
[the remark this answers is not in the document]
Comment (YuRihl) [reply src/auth/auth.service.ts:59]: Done

File: src/auth/auth.controller.ts [hunk 8]
```diff
@@ -1,14 +1,27 @@
...
-import { UserAuthGuard } from './user-auth.guard';
+import { RegisterDto } from './dto';
+import { LoginDto } from './dto/login.dto';
 
 @Controller()
 export class AuthController {
   constructor(private readonly authService: AuthService) {}
 
-  @UseGuards(UserAuthGuard)
+  //@UseGuards(UserAuthGuard)
```
[the remark this answers is not in the document]
Comment (YuRihl) [reply src/auth/auth.controller.ts:12]: Done

File: .env [hunk 9]
```diff
@@ -15,3 +15,6 @@ DB_PASSWORD=dbyuripass
 PGADMIN_EMAIL=goliney2001g@gmail.com
 PGADMIN_PASSWORD=dbyuripass
 PGADMIN_PORT=5050
+
+# JWT
+JWT_SECRET_KEY=authkey
```
[the remark this answers is not in the document]
Comment (YuRihl) [reply .env:20]: I didn't get it, you mean to create additional .env.example file for storing JWT key to make config more modular?

**Pull request merged**